src/transformers/models/encdecgemma2/modular_encdecgemma2.py

```diff
@@ -0,0 +1,1637 @@
+            [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+@add_start_docstrings(
+    "The bare EncdecGemma2 Model outputting raw hidden-states without any specific head on top.",
+    ENCDECGEMMA2_START_DOCSTRING,
+)
+class EncdecGemma2PreTrainedModel(Gemma2PreTrainedModel):
+    config_class = EncdecGemma2Config
+    base_model_prefix = "model"
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["EncdecGemma2Block"]
+
+    def _init_weights(self, module):
+        # TODO: support intialization for encoders and decoders separately(?)
+        std = self.config.initializer_range
+        if isinstance(module, nn.Linear):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+        elif isinstance(module, EncdecGemma2RMSNorm):
+            module.weight.data.fill_(1.0)
+        elif isinstance(module, EncdecGemma2ClassificationHead):
+            scale = module.out_proj.weight.shape[0] ** -0.5
+            module.out_proj.weight.data.normal_(mean=0.0, std=std * scale)
+            if hasattr(module.out_proj, "bias") and module.out_proj.bias is not None:
+                module.out_proj.bias.data.zero_()
+        elif isinstance(module, EncdecGemma2LMHead):
+            if not self.config.tie_word_embeddings:
+                scale = module.out_proj.weight.shape[0] ** -0.5
+                module.out_proj.weight.data.normal_(mean=0.0, std=std * scale)
+
+    def _shift_right(self, input_ids):
+        """
+        Shifts input_ids to the right, prepends the decoder_start_token_id, and handles
+        pad_token_id replacement for labels that were -100.
+        This is a common preparation step for decoder inputs in sequence-to-sequence models.
+        """
+        decoder_start_token_id = self.config.decoder.bos_token_id
+        pad_token_id = self.config.decoder.pad_token_id
+
+        if decoder_start_token_id is None:
+            raise ValueError(
+                "self.model.config.decoder.bos_token_id has to be defined. "
+            )
+
+        # shift inputs to the right
+        shifted_input_ids = input_ids.new_zeros(input_ids.shape)
+        shifted_input_ids[..., 1:] = input_ids[..., :-1].clone()
+        shifted_input_ids[..., 0] = decoder_start_token_id
+
+        if pad_token_id is None:
+            raise ValueError("self.model.config.decoder.pad_token_id has to be defined.")
+        
+        # Is this T5 specific?
+        # replace possible -100 values in labels by `pad_token_id`
+        shifted_input_ids.masked_fill_(shifted_input_ids == -100, pad_token_id)
+
+        return shifted_input_ids
+
+
+class EncdecGemma2Stack(EncdecGemma2PreTrainedModel):
```

Comment:
not a fan of stacks, let's split encoder and decoder! I know t5 has that but it's not a good precedent!